src/main/java/com/shapesecurity/shift/es2016/template/Template.java

```diff
@@ -0,0 +1,207 @@
+
+import com.shapesecurity.functional.F;
+import com.shapesecurity.functional.F2;
+import com.shapesecurity.functional.Pair;
+import com.shapesecurity.functional.data.ImmutableList;
+import com.shapesecurity.functional.data.Maybe;
+import com.shapesecurity.shift.es2016.ast.BindingIdentifier;
+import com.shapesecurity.shift.es2016.ast.Node;
+import com.shapesecurity.shift.es2016.ast.Program;
+import com.shapesecurity.shift.es2016.ast.Script;
+import com.shapesecurity.shift.es2016.parser.JsError;
+import com.shapesecurity.shift.es2016.parser.ParserWithLocation;
+import com.shapesecurity.shift.es2016.parser.SourceSpan;
+import com.shapesecurity.shift.es2016.reducer.Director;
+import com.shapesecurity.shift.es2016.reducer.Flattener;
+import com.shapesecurity.shift.es2016.reducer.ReconstructingReducer;
+import com.shapesecurity.shift.es2016.reducer.Reducer;
+import com.shapesecurity.shift.es2016.reducer.WrappedReducer;
+import com.shapesecurity.shift.es2016.utils.WithLocation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Template {
+	public static class NodeInfo {
+		public final String name;
+		public final Node node;
+		public final ParserWithLocation.Comment comment;
+
+		NodeInfo(String name, Node node, ParserWithLocation.Comment comment) {
+			this.name = name;
+			this.node = node;
+			this.comment = comment;
+		}
+	}
+
+	static class Marker {
+		final String name;
+		final Maybe<Class<? extends Node>> type;
+		final int start;
+		final ParserWithLocation.Comment comment;
+
+		Marker(String name, Maybe<Class<? extends Node>> type, int start, ParserWithLocation.Comment comment) {
+			this.name = name;
+			this.type = type;
+			this.start = start;
+			this.comment = comment;
+		}
+	}
+
+	public static Program replace(Program tree, F2<Node, Node, Maybe<Node>> getReplacement) {
+		Reducer<Node> replacer = new WrappedReducer<>(
+			(originalNode, newNode) -> getReplacement.apply(newNode, originalNode).orJust(newNode),
+			new ReconstructingReducer() // TODO LazyCloneReducer
+		);
+		return (Program) Director.reduceProgram(replacer, tree);
+	}
+
+	static final Pattern commentRegex = Pattern.compile("^# ([^#]+) (?:# ([^#]+) )?#$");
+	public static Maybe<Pair<String, Maybe<Class<? extends Node>>>> parseComment(String text) {
+		Matcher matcher = commentRegex.matcher(text);
+		if (!matcher.matches()) {
+			return Maybe.empty();
+		}
+
+		String name = matcher.group(1);
+		String typeName = matcher.group(2);
+
+		if (typeName != null) {
+			try {
+				Class type = Class.forName("com.shapesecurity.shift.es2016.ast." + typeName); // This is kind of awful, but... it does work...
```

Comment:
I mean, you could manually define a mapping, but this is fine. At least check that there's no `.`s or other trickery in `typeName`.